Philips.CodeAnalysis.MaintainabilityAnalyzers/Readability/AvoidMultipleLambdasOnSingleLineAnalyzer.cs

```diff
@@ -0,0 +1,74 @@
+﻿// © 2023 Koninklijke Philips N.V. See License.md in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Philips.CodeAnalysis.Common;
+
+namespace Philips.CodeAnalysis.MaintainabilityAnalyzers.Readability
+{
+	[DiagnosticAnalyzer(LanguageNames.CSharp)]
+	public class AvoidMultipleLambdasOnSingleLineAnalyzer : DiagnosticAnalyzer
+	{
+
+		private const string Title = @"Avoid putting multiple Lambda expressions on the same line";
+		public const string MessageFormat = @"Avoid putting multiple Lambda expressions on the same line";
+		private const string Description = @"Avoid putting multiple Lambda expressions on the same line. To improve readability, split them into separate lines.";
+		private const string Category = Categories.Readability;
+
+		private static readonly DiagnosticDescriptor Rule = new(Helper.ToDiagnosticId(DiagnosticIds.AvoidMultipleLambdasOnSingleLine), Title,
+			MessageFormat, Category, DiagnosticSeverity.Error, isEnabledByDefault: true, description: Description);
+
+
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+		public override void Initialize(AnalysisContext context)
+		{
+			context.EnableConcurrentExecution();
+			context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+			
+			context.RegisterSyntaxNodeAction(Analyze, SyntaxKind.ClassDeclaration);
```

Comment:
These feels like it's hitting too much. Shouldn't this be LambdaExpressionSyntax? And then use the Span to find any other lambdas?

Reply:
Moving from one Lambda to another is complicated, as they are not in a flat hierarchical relationship in the SyntaxTree.

For me, the Visitor design nicely sorts them per line.

Reply:
I guess I'm still confused... The implementation, if I understand correctly, walks the entire code base, for just this 1 analyzer. Could we not trigger only on LambaExpressionSyntax? From there, if we want to walk the syntax tree, we can first just jump up to the MethodDeclarationSyntax ancestor? (This would suffer from multiple visits of the same method when there are multiple lambas in that method, but that's solvable.) Wouldn't this give ~75-99% performance improvement assuming only a percentage of methods use lambdas?

Reply:
Ok, now I get your original remark.

I now trigger on Lambda, walk the tree up to the TypeDeclaration and continue as-is.
Put a protection in to avoid Analyzing the same type more then once.